keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1475,225 +1382,400 @@ prevented. Deleting the entire cluster also leaves resources allocated outside
-    metav1.TypeMeta
-    // Standard object metadata
-    metav1.ObjectMeta
 
-    // If this object was created from some other resource, then this links
-    // back to that resource. This field is used to find the in-tree representation
-    // of the claim parameters when the parameter reference of the claim refers
-    // to some unknown type.
-    GeneratedFrom *ResourceClaimParametersReference
+// Exactly one field must be set.
+type Selector struct {
+    // This CEL expression must evaluate to true if a
+    // device is suitable. This covers qualitative aspects of
+    // device selection.
+    //
+    // The language is as defined in
+    // https://kubernetes.io/docs/reference/using-api/cel/
+    // with several additions that are specific to device selectors.
+    //
+    // Attributes of a device are made available through a `device.attributes` map.
+    // The value of each entry varies, depending on the attribute
+    // that is being looked up.
+    //
+    // Unknown keys trigger a runtime error.
+    //
+    // The `device.driverName` string variable can be used to check for a specific
+    // driver explicitly in a filter that is meant to work for devices from
+    // different vendors. It is provided by Kubernetes and matches the
+    // `driverName` from the ResourceSlice which provides the device.
+    //
+    // The CEL expression is applied to *all* available devices from any driver.
+    // The expression has to check for existence of an attribute when it is not
+    // certain that it is provided because runtime errors are not automatically
+    // treated as "don't select device". Instead, device selection fails completely
+    // and reports the error.
+    //
+    // Some examples:
+    //    "memory.dra.example.com" in device.attributes && # Is the attribute available?
+    //       device.attributes["memory.dra.example.com"].isGreaterThan(quantity("1Gi")) # >= 1Gi
+    //    device.driverName == "dra.example.com" # any device from that driver
+    //
+    // +optional
+    Expression *string
+}
+```
 
-    // Shareable indicates whether the allocated claim is meant to be shareable
-    // by multiple consumers at the same time.
-    Shareable bool
+```
+<<[UNRESOLVED @pohly ]>>
+
+The `device.` prefix is included in the CEL environment for the reasons given in
+https://github.com/kubernetes-sigs/wg-device-management/issues/23#issuecomment-2160044359.
+
+Whether we really do it like that will be discussed further in that issue and/or during
+the implementation phase. Dropping it is not a conceptual change.
+
+<<[/UNRESOLVED]>>
+```
+
+```go
+
+// Exactly one field must be set.
+type Amount struct {
+    // All, if set, asks for all devices matching the selectors.
+    // Allocation fails if not all of them are available, unless admin
+    // access is requested. Admin access is granted also for
+    // devices which are in use.
+    //
+    // The default if `all` and `count` are unset is to allocate one device.
```

Comment:
Same trap as above - you can't have nothing mean something.  If we ever wanted to add a `range` option, but the client was back-rev, it would see `all` and `count` as unset and assume 1, which is wrong.

You can't even get away with saying that the apiserver will default count to 1 if no other field is set.  If you have a modern client and a back-rev apiserver, and the client passes `range` (which the apiserver doesn't know about), the default will kick in, corrupting the client's intention.

AFAICT your choices are:

1) never add to this one-of and rely on "nothing set" = default

2) require exactly 1 field, which means trivial use-cases have to say `count: 1`

3) add a discriminator, if the discriminator is not set it gets defaulted to "Exact".  Any other mode REQUIRES the discriminator be set.

We don't often do the discriminator thing in k8s.  We have made the mistake several times and found ourselves in case 1 (e.g. NetworkPolicy).  We usually can make case 2 work.  In this case I think you might want 3.

Simple case: say nothing, get default `CountMode: Exact` and `Count: 1`

If I want 2: set `CountMode: Exact` and `Count: 2`

If I want all: set `CountMode: All`

If I want a range: set `CountMode: Range` and `Range: { min: 1, max: 4 }`

Reply:
Ahh. That makes sense. I think the discriminator is probably what we want here.

Reply:
The `Amount` struct is referenced via a pointer in `RequestDetail`. Therefore if we add a `range`, the back-rev client will see a pointer to an empty struct, which is safe.

The "if `all` and `count` are unset" is how the users sees it and therefore how the documentation is written. What the apiserver and client see is a nil `*Amount` pointer.

Reply:
So, you are saying, "nil" means 1, but "empty" means "there must be something I don't now"; that is, we can differentiate as it is written, no need to change?

Reply:
I don't think so.  It's too clever and depends on the content-type and language/library details.

E.g. we support JSON as a transport.  A forward-rev sender who sends `range` will unmarshall as a `nil` Amount in the receiver:  https://go.dev/play/p/eQGyV5_UsIV

Also what does it mean if I pass `all: false` and nothing else?

Reply:
> A forward-rev sender who sends range will unmarshall as a nil Amount in the receiver: https://go.dev/play/p/eQGyV5_UsIV

Ack. That is where it breaks down.

> 3. add a discriminator, if the discriminator is not set it gets defaulted to "Exact". Any other mode REQUIRES the discriminator be set.
>
> Simple case: say nothing, get default CountMode: Exact and Count: 1

Does that literally have to be server-side defaulting or can we say that nil CountMode is the same as "Exact" and in that case, nil Count is the same as 1?

It's a bit odd to send `CountMode: Range` and `Range: { min: 1, max: 4 }` and then see a default `Count: 1` - okay, so we could only set that default for `CountMode: Exact`... hmm.

Reply:
defaults:
```
if x.CountMode == nil {
  x.CountMode == Exact
}
if numFieldsSet(x) == 0 {
  x.Count = 1
}
```

validation:
```
if x.CountMode == nil {
  // error: required
}
switch x.CountMode {
case Exact:
    if Count <= 0 {
        // error: GTZ
    }
case All:
  // do nothing
case Range:
  if x.Min <= 0 {
    // error: GTZ
  }
  if x.Max > 128 {
    // error?
  }
  if x.Min > x.Max {
    // error
  }
```

You could MAYBE cross-check (e.g. Range is not set when mode=Exact, Range and Count are not set when mode=All), too.

Reply:
I'm not sure whether I like the `CountMode` discriminator. I'm going with an extra level and a pointer to one-of struct for now in the implementation (`Amount: Count: 5`, `Amount: All: true`).

Reply:
We can fight in API review, if you prefer.  

Adding a nesting COULD solve the "nothing means something" problem, but only if the `Amount` field is itself a pointer (despite not really being optional):

https://go.dev/play/p/ig-JnD0jcn9
  vs
https://go.dev/play/p/EW39t7lCqIG

This feels like playing tricks with unmarshalling, and I worry about clients or protocols which might not be so clever.    Also, I really dislike `All bool`.

Reply:
For the record, I really dislike `All bool` as well, but don't see a way around it unless we define a new type, which can either be an integer (represented as a string) or the special keyword "all".

Then you can have e.g. `count: "1"`, `count: "2"`, or `count: "all"`.

Reply:
This count "type" would have to be stored as a string to avoid issues with round-tripping. We do the same for semver.

The problem with "extend it" is that if version X+1 allows `count: "1-3"` and stores that, then downgrading to X where that wasn't allowed yet leads to an object which isn't valid. That causes all kinds of nastiness, I think even not being able to delete it.

Reply:
Let's continue here, but I also added a cross-reference in https://github.com/kubernetes/kubernetes/pull/125488/commits/d4b048c8b33cac6c045aa5d029aab4e405671fce#r1647009797.

@thockin: if we allow `countMode: exact` and `countMode: all` in 1.31 and validation rejects everything else, how do we add `countMode: range` later? At first glance it has the same problem with downgrades. Or does validation perhaps only get applied on creation? On updates, we already need to check for "spec not modified", so if that passes, perhaps we can simply skip the entire spec validation?

Reply:
If we do that, then the `count: all` suggested by @klueska may be the simplest approach.

Reply:
> This count "type" would have to be stored as a string to avoid issues with round-tripping.

But as we probably want to allow `count: 1`, an [IntOrString](https://pkg.go.dev/github.com/kubernetes/apimachinery/pkg/util/intstr#IntOrString) is a better choice.

Reply:
IntOrString is generally seen as a bad idea.  JSON supports it, but protobuf (and any other reasonable transport) does not.  We should not repeat that, I think.

> if we allow countMode: exact and countMode: all in 1.31 and validation rejects everything else, how do we add countMode: range later? 

API changes require an alpha.  Assuming that we are successful in getting DRA to beat in 31:

1.31: we have only the main DRA gate
DRA gate (alpha) off: NA, not usable  <-- default
DRA gate (alpha) on: would see "countMode: range" and reject it

1.32: we add a new gate for DRA count ranges
DRA gate (beta) off: NA, not usable
DRA gate (beta) on + range gate (alpha) off: would see "countMode: range" and reject it  <-- default
DRA gate (beta) on + range gate (alpha) on: would see "countMode: range" and allow it

1.33: we have 2 gates
DRA gate (beta) off: NA, not usable
DRA gate (beta) on + range gate (beta) off: would see "countMode: range" and reject it
DRA gate (beta) on + range gate (beta) on: would see "countMode: range" and allow it  <-- default

Any back-rev scheduler or autoscaler would say "I don't know what to do with that" and reject it.

The important piece is that API gate-checks are always "gate_enabled || already_in_use".  In 1.32 you would have to manually enable the range gate (alpha), which waives rollback guarantees.  In 1.33 the range gate (beta) would be on by default.  If you used a range in 1.33, then rolled back to 1.32, it meets the "already_in_use" criterion. and would not fail validation.

Embedding the data into a string could ALSO satisfy this pattern, but we rarely, if ever, do formatted strings this way.  The few places we have done it tend to be awkward (e.g. container image tags forcing consumers to parse for `:latest`).  This would either abuse IntOrString (bad) or carry numbers as strings (bad, and yes I hate Quantity for this) and require a grammar and parsing (bad).

Or you can just require that clients pass exactly one option, making `count: 1` explicit (no default).

Reply:
So the `already_in_use` part is what makes validation not break when downgrading or disabling feature gates. That applies to adding new enum values as well as strings which need be parsed. But `amount: "1"` (now) and `amount: "1-2"` (future extension) requires parsing, so that's out.

I think that leaves @thockin's [original proposal](https://github.com/kubernetes/enhancements/pull/4709#discussion_r1640552271) in this thread as the most viable approach (discriminator with additional fields). I'll keep the current simplistic one-of struct in my implementation PR for now, but will switch to the discriminator approach soon.

Reply:
Thanks for working thru it.  I am trying to collect information to write a clearer API convention on this specifically.

Reply:
Included in https://github.com/kubernetes/kubernetes/pull/125488.